crates/turborepo-lockfiles/src/npm.rs

```diff
@@ -0,0 +1,365 @@
+    }
+}
+
+impl NpmLockfile {
+    pub fn load(content: &[u8]) -> Result<Self, Error> {
+        let lockfile: NpmLockfile = serde_json::from_slice(content)?;
+
+        // We don't support lockfiles without 'packages' as older versions
+        // required reading through the contents of node_modules in order
+        // to resolve dependencies.
+        // See https://github.com/npm/cli/blob/9609e9eed87c735f0319ac0af265f4d406cbf800/workspaces/arborist/lib/shrinkwrap.js#L674
+        if lockfile.lockfile_version <= 1
+            || (lockfile.packages.is_empty() && !lockfile.dependencies.is_empty())
+        {
+            Err(Error::UnsupportedNpmVersion)
+        } else {
+            Ok(lockfile)
+        }
+    }
+
+    fn get_package(&self, package: impl AsRef<str>) -> Result<&NpmPackage, Error> {
+        let pkg_str = package.as_ref();
+        self.packages
+            .get(pkg_str)
+            .ok_or_else(|| Error::MissingPackage(pkg_str.to_string()))
+    }
+
+    pub fn subgraph(
+        &self,
+        workspace_packages: &[String],
+        packages: &[String],
+    ) -> Result<Self, Error> {
+        let mut pruned_packages = HashMap::with_capacity(packages.len());
+        for pkg_key in packages {
+            let pkg = self.get_package(pkg_key)?;
+            pruned_packages.insert(pkg_key.to_string(), pkg.clone());
+        }
+        if let Some(root) = self.packages.get("") {
+            pruned_packages.insert("".into(), root.clone());
+        }
+        for workspace in workspace_packages {
+            let pkg = self.get_package(workspace)?;
+            pruned_packages.insert(workspace.to_string(), pkg.clone());
+
+            for (key, entry) in &self.packages {
+                if entry.resolved.as_deref() == Some(workspace) {
+                    pruned_packages.insert(key.clone(), entry.clone());
+                    break;
+                }
+            }
+        }
+        Ok(Self {
+            lockfile_version: 3,
+            packages: pruned_packages,
+            dependencies: HashMap::default(),
+            other: self.other.clone(),
+        })
+    }
+
+    fn possible_npm_deps(key: &str, dep: &str) -> Vec<String> {
+        let mut possible_deps = vec![format!("{key}/node_modules/{dep}")];
+
+        let mut curr = Some(key);
+        while let Some(key) = curr {
+            let next = Self::npm_path_parent(key);
+            possible_deps.push(format!("{}node_modules/{}", next.unwrap_or(""), dep));
+            curr = next;
+        }
+
+        possible_deps
+    }
+
+    fn npm_path_parent(key: &str) -> Option<&str> {
+        key.rsplit_once("node_modules/")
+            .map(|(first, _)| first)
+            .and_then(|parent| {
+                if parent.is_empty() {
+                    None
+                } else {
+                    Some(parent)
+                }
+            })
+    }
+}
+
+impl NpmPackage {
+    pub fn dep_keys(&self) -> impl Iterator<Item = &String> {
```

Comment:
Love this :)

Reply:
I've been missing my iterators while I've been in Go land